streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -389,6 +391,18 @@ public void close() {
         metrics.removeAllSensors();
     }
 
+    void closeProducer() {
+        if (exactlyOnceEnabled) {
+            try {
+                recordCollector.close();
+            } catch (final Throwable e) {
+                log.error("{} Failed to close producer: ", logPrefix, e);
+            }
+        } else {
+            throw new IllegalStateException("Tasks should only close producers if exactly-once semantics is enabled.");
```

Comment:
I think we are making things more coupled than they need to be. Why don't we introduce the notion of `StreamsTask.close`. This method should always be called. In `RecordCollectorImpl`, we take a boolean that indicates whether it owns the Producer or not. If it does, it closes the producer in the `close` method. If not, it doesn't. This way `StreamsTask` doesn't even need to care about the `exactlyOnce` setting.